crates/ruff/src/rules/flake8_pyi/rules/duplicate_union_member.rs

```diff
@@ -0,0 +1,79 @@
+use rustc_hash::FxHashSet;
+use rustpython_parser::ast::{Expr, ExprKind, Operator};
+
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::comparable::ComparableExpr;
+use ruff_python_ast::helpers::unparse_expr;
+use ruff_python_ast::types::Range;
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+#[violation]
+pub struct DuplicateUnionMember {
+    pub duplicate_name: String,
+}
+
+impl AlwaysAutofixableViolation for DuplicateUnionMember {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Duplicate union member `{}`", self.duplicate_name)
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Remove duplicate union member `{}`", self.duplicate_name)
+    }
+}
+
+/// PYI016
+pub fn duplicate_union_member(checker: &mut Checker, expr: &Expr) {
+    let mut seen_nodes = FxHashSet::default();
+    traverse_union(&mut seen_nodes, checker, expr, None);
+}
+
+fn traverse_union<'a>(
+    seen_nodes: &mut FxHashSet<ComparableExpr<'a>>,
+    checker: &mut Checker,
+    expr: &'a Expr,
+    previous: Option<&'a Expr>,
+) {
+    // The union data structure usually looks like this:
+    //  a | b | c -> (a | b) | c
+    //
+    // However, parenthesized expressions can coerce it into any structure:
+    //  a | (b | c)
+    //
+    // So we have to traverse both branches in order (left, then right), to report duplicates
+    // in the order they appear in the source code.
+    if let ExprKind::BinOp {
+        op: Operator::BitOr,
+        left,
+        right,
+    } = &expr.node
+    {
+        // Traverse left subtree, then the right subtree, propagating the previous node.
+        traverse_union(seen_nodes, checker, left, previous);
+        traverse_union(seen_nodes, checker, right, Some(left));
+    }
+
+    // If we've already seen this union member, raise a violation.
+    if !seen_nodes.insert(expr.into()) {
+        let mut diagnostic = Diagnostic::new(
+            DuplicateUnionMember {
+                duplicate_name: unparse_expr(expr, checker.stylist),
+            },
+            Range::from(expr),
+        );
+        if checker.patch(diagnostic.kind.rule()) {
+            // Delete the "|" character as well as the duplicate value. We structure this as
+            // a deletion from the end of `previous` to the end of `expr`.
+            diagnostic.set_fix(Edit::deletion(
+                // SAFETY: impossible to have a duplicate without a `previous` node.
+                previous.unwrap().end_location.unwrap(),
+                expr.end_location.unwrap(),
+            ));
```

Comment:
I think this can lead to syntax errors in the presence of parentheses. E.g., if you try to autofix this, we throw a syntax error:

```py
field2: (str | int) | str  # PYI016 Duplicate name in union
```

I'm guessing the issue is that the previous `end_location` won't include the parenthesis -- so we change to:

```py
field2: (str | int  # PYI016 Duplicate name in union
```

(I'm guessing here, I haven't looked at the erroneous source code.)

As an alternative strategy, we could use `unparse_expr`, and re-create the source `expr`, but with the duplicate expression removed. It'll be guaranteed to be correct and valid syntax, but we will lose trivia (comments, optional user-added parentheses).

Reply:
Ok, I added a couple test cases to the fixture for these brackets, which work as I'd expect:

```python
field9: int | (int | str)  # PYI016 Duplicate union member `int`
field10: (str | int) | str  # PYI016 Duplicate union member `str`
# Gets auto fixed to
field9: int | (str)  # PYI016 Duplicate union member `int`
field10: str | int  # PYI016 Duplicate union member `str`
```

Reply:
Cool, this looks good to me!

Reply:
Yeah, the approach here makes sense. I guess I was suggesting taking the top-level expression (the thing we pass into `duplicate_union_member`) and recursively filtering to find the duplicated child, but what you settled on here seems better.